File: backend/legacy_ts_archive/src/services/ProjectService.ts
import { DbProject, DbProjectLink } from '../models/db.js';
import { ProjectRepository } from '../repositories/ProjectRepository.js';

export interface CreateProjectDTO {
  name: string;
  shortDescription: string;
  detailedDescription?: string;
  problemStatement?: string;
  solution?: string;
  targetUsers?: string;
  category?: string;
  eventName: string;
  eventOrganizer?: string;
  eventType?: string;
  eventDate?: string;
  eventDescription?: string;
  techFrontend?: string;
  techBackend?: string;
  techDatabase?: string;
  techAi?: string;
  techApis?: string;
  techHosting?: string;
  techOther?: string;
  feedbackGoal?: string;
  isAuthorized: boolean;
  links?: Array<{ linkType: string; url: string; label?: string }>;
}

export class ProjectService {
  public static async createProject(userId: string, data: CreateProjectDTO): Promise<{ project: DbProject; links: DbProjectLink[] }> {
    if (!data.isAuthorized) {
      throw new Error('Project authorization confirmation is required before proceeding.');
    }

    if (!data.name || !data.shortDescription || !data.eventName) {
      throw new Error('Project name, short description, and event name are required.');
    }

    const project = await ProjectRepository.create({
      userId,
      name: data.name.trim(),
      shortDescription: data.shortDescription.trim(),
      detailedDescription: data.detailedDescription?.trim(),
      problemStatement: data.problemStatement?.trim(),
      solution: data.solution?.trim(),
      targetUsers: data.targetUsers?.trim(),
      category: data.category || 'General',
      eventName: data.eventName.trim(),
      eventOrganizer: data.eventOrganizer?.trim(),
      eventType: data.eventType || 'Hackathon',
      eventDate: data.eventDate,
      eventDescription: data.eventDescription,
      techFrontend: data.techFrontend,
      techBackend: data.techBackend,
      techDatabase: data.techDatabase,
      techAi: data.techAi,
      techApis: data.techApis,
      techHosting: data.techHosting,
      techOther: data.techOther,
      feedbackGoal: data.feedbackGoal || 'General Feedback',
      authorizedAt: new Date()
    });

    const createdLinks: DbProjectLink[] = [];
    if (data.links && Array.isArray(data.links)) {
      for (const link of data.links) {
        if (link.url && link.linkType) {
          const l = await ProjectRepository.addLink(project.id, link.linkType, link.url, link.label);
          createdLinks.push(l);
        }
      }
    }

    return { project, links: createdLinks };
  }

  public static async getProject(projectId: string, userId: string): Promise<{ project: DbProject; links: DbProjectLink[] }> {
    const project = await ProjectRepository.findByIdAndOwner(projectId, userId);
    if (!project) {
      throw new Error('Project not found or access denied (IDOR protection)');
    }
    const links = await ProjectRepository.getLinks(projectId);
    return { project, links };
  }

  public static async listProjects(userId: string): Promise<DbProject[]> {
    return await ProjectRepository.listByUser(userId);
  }

  public static async updateProject(projectId: string, userId: string, updates: Partial<CreateProjectDTO>): Promise<DbProject> {
    const updated = await ProjectRepository.update(projectId, userId, updates as any);
    if (!updated) {
      throw new Error('Project not found or access denied');
    }
    return updated;
  }

  public static async deleteProject(projectId: string, userId: string): Promise<void> {
    const success = await ProjectRepository.softDelete(projectId, userId);
    if (!success) {
      throw new Error('Project not found or access denied');
    }
  }
}

File: backend/app/services/project_service.py
"""Project Management Service with Strict Ownership Enforcement."""

from datetime import datetime, timezone
import uuid

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.db.models.project import Project, ProjectLink
from app.schemas.project import ProjectCreate, ProjectLinkCreate, ProjectUpdate


class ProjectService:
    @classmethod
    async def create_project(cls, db: AsyncSession, user_id: uuid.UUID, data: ProjectCreate) -> Project:
        now = datetime.now(timezone.utc)
        authorized_at = now if data.authorize_audit else None

        project = Project(
            user_id=user_id,
            name=data.name.strip(),
            short_description=data.short_description.strip(),
            detailed_description=data.detailed_description.strip() if data.detailed_description else None,
            problem_statement=data.problem_statement.strip() if data.problem_statement else None,
            solution=data.solution.strip() if data.solution else None,
            target_users=data.target_users.strip() if data.target_users else None,
            category=data.category,
            event_name=data.event_name.strip(),
            event_organizer=data.event_organizer.strip() if data.event_organizer else None,
            event_type=data.event_type,
            event_date=data.event_date,
            event_description=data.event_description,
            tech_frontend=data.tech_frontend,
            tech_backend=data.tech_backend,
            tech_database=data.tech_database,
            tech_ai=data.tech_ai,
            tech_apis=data.tech_apis,
            tech_hosting=data.tech_hosting,
            tech_other=data.tech_other,
            feedback_goal=data.feedback_goal,
            authorized_at=authorized_at,
        )
        db.add(project)
        await db.flush()

        for link_in in data.links:
            link = ProjectLink(
                project_id=project.id,
                link_type=link_in.link_type,
                url=link_in.url.strip(),
                label=link_in.label.strip() if link_in.label else None,
            )
            db.add(link)

        await db.commit()

        # Reload with links
        stmt = (
            select(Project)
            .where(Project.id == project.id)
            .options(selectinload(Project.links))
        )
        return await db.scalar(stmt)

    @classmethod
    async def list_projects(cls, db: AsyncSession, user_id: uuid.UUID) -> list[Project]:
        stmt = (
            select(Project)
            .where(Project.user_id == user_id, Project.is_active.is_(True))
            .options(selectinload(Project.links))
            .order_by(Project.created_at.desc())
        )
        res = await db.scalars(stmt)
        return list(res.all())

    @classmethod
    async def get_project(cls, db: AsyncSession, project_id: uuid.UUID, user_id: uuid.UUID) -> Project:
        stmt = (
            select(Project)
            .where(Project.id == project_id, Project.user_id == user_id)
            .options(selectinload(Project.links))
        )
        project = await db.scalar(stmt)
        if not project:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Project not found or access denied.",
            )
        return project

    @classmethod
    async def update_project(
        cls, db: AsyncSession, project_id: uuid.UUID, user_id: uuid.UUID, data: ProjectUpdate
    ) -> Project:
        project = await cls.get_project(db, project_id, user_id)

        update_dict = data.model_dump(exclude_unset=True)
        if "authorize_audit" in update_dict:
            authorize = update_dict.pop("authorize_audit")
            project.authorized_at = datetime.now(timezone.utc) if authorize else None

        for field, value in update_dict.items():
            setattr(project, field, value)

        await db.commit()
        await db.refresh(project)
        return project

    @classmethod
    async def delete_project(cls, db: AsyncSession, project_id: uuid.UUID, user_id: uuid.UUID) -> None:
        project = await cls.get_project(db, project_id, user_id)
        await db.delete(project)
        await db.commit()

    @classmethod
    async def add_link(
        cls, db: AsyncSession, project_id: uuid.UUID, user_id: uuid.UUID, link_data: ProjectLinkCreate
    ) -> ProjectLink:
        # verify ownership
        await cls.get_project(db, project_id, user_id)
        link = ProjectLink(
            project_id=project_id,
            link_type=link_data.link_type,
            url=link_data.url.strip(),
            label=link_data.label.strip() if link_data.label else None,
        )
        db.add(link)
        await db.commit()
        await db.refresh(link)
        return link
